### PR TITLE
AnimationSprite class 1st try ^^

### PR DESCRIPTION
**Pull request by Romuald78**
It is possible to create time based sprite animations :
- flip H and V
- apply color filter for each animation
- sets a limited number of animation loops
- set back and forth animation
- resize final animation at loading
- use several animations for each facing direction

needs to be tested more but it seems to be pretty functional ;)
Enjoy !!

### CODE REVIEW

File: arcade/sprite.py [hunk 1]
```diff
@@ -1054,6 +1054,243 @@ def update_animation(self, delta_time: float = 1/60):
...
+        self._current_animation_name = None
+        # Current displayed texture
+        self.cur_texture_index = 0
+        # Set elapsed duration (used to know if we have to stop the animation)
+        self._elapsed_duration = 0
+        # Set play/pause flag
+        self._playing = True
+
+
+    def addAnimation(self,
```
Comment (einarf) [inline arcade/sprite.py]: Should probably use snake case for all method instead of camel case

File: arcade/sprite.py [hunk 2]
```diff
@@ -1054,6 +1054,243 @@ def update_animation(self, delta_time: float = 1/60):
...
+
+            # In case of backAndForth
+            if frameIdx >= nbFrames:
+                frameIdx = nbFrames2 - frameIdx
+
+            # set current texture index and texture from the parent class
+            self.cur_texture_index = frameIdx
+            self.set_texture(self.cur_texture_index)
+
+    def pauseAnimation(self):
```
Comment (einarf) [inline arcade/sprite.py]: use snake case

File: arcade/sprite.py [hunk 3]
```diff
@@ -1054,6 +1054,243 @@ def update_animation(self, delta_time: float = 1/60):
...
+        """
+        Pauses the current animation. It does not rewind it.
+        :return: None
+        """
+        self._playing = False
+
+    def resumeAnimation(self):
```
Comment (einarf) [inline arcade/sprite.py]: use snake case

File: arcade/sprite.py [hunk 4]
```diff
@@ -1054,6 +1054,243 @@ def update_animation(self, delta_time: float = 1/60):
...
+        """
+        Resumes the current animation. It does not rewind it.
+        :return: None
+        """
+        self._playing = True
+
+    def rewindAnimation(self):
```
Comment (einarf) [inline arcade/sprite.py]: use snake case

File: arcade/sprite.py [hunk 5]
```diff
@@ -1054,6 +1054,243 @@ def update_animation(self, delta_time: float = 1/60):
...
+        """
+        Just rewinds the current animation to the first frame. It does not change the play/stop flag.
+        :return:
+        """
+        self._elapsed_duration = 0
+
+    def playAnimation(self):
```
Comment (einarf) [inline arcade/sprite.py]: use snake case

File: arcade/sprite.py [hunk 6]
```diff
@@ -1054,6 +1054,243 @@ def update_animation(self, delta_time: float = 1/60):
...
+        """
+        Rewinds and Plays the current animation.
+        :return: None
+        """
+        self.rewindAnimation()
+        self.resumeAnimation()
+
+    def stopAnimation(self):
```
Comment (einarf) [inline arcade/sprite.py]: use snake case

File: arcade/sprite.py [hunk 7]
```diff
@@ -1054,6 +1054,243 @@ def update_animation(self, delta_time: float = 1/60):
...
+
+    """"
+    New class for Sprite animation. (trial)
+    """
+
+    # static integer for unique texture ids
+    _uniqueID = 1
+
+    @staticmethod
+    def _getNextID():
```
Comment (einarf) [inline arcade/sprite.py]: maybe something like `_next_id()`

File: arcade/sprite.py [hunk 8]
```diff
@@ -1054,6 +1054,243 @@ def update_animation(self, delta_time: float = 1/60):
...
 
+
+class AnimatedSprite(Sprite):
+
+    """"
+    New class for Sprite animation. (trial)
+    """
+
+    # static integer for unique texture ids
+    _uniqueID = 1
```
Comment (einarf) [inline arcade/sprite.py]: snake case

**Comment by einarf** [comment]
There are many places were camelCase is used here. We like to use snake case in python. 

If this new class is something arcade needs I'll leave up to @pvcraven 😄 (I mainly work on the lower level part of the library)

**Comment by Romuald78** [comment]
I didnt checked AnimatedTimeBaseSprite class, but is seems the class i provided is something similar/redundant/reworked 
I tried to use the texture list from Sprite class, and the state field for the 4 facing directions, so i really hope it can be useful for arcade 😉
Anyway i ll update according to the naming conventions

**Comment by Romuald78** [comment]
Reminder : add a method is_finished() in order to knowif the animation of a Sprite is over (in case game logic wants to execute some code at this moment)
Reminder2 : ads a method get_percent or similar to get the progression of the animation in case the game logic wants to deal with it
Reminder3 : add the way to select specifically the texture index, get the number of current index etc... pairing select texture and stop can lead to have a fixed sprite with multi displays

**Comment by Romuald78** [comment]
@pvcraven i guess you have a busy planning and limited time to offer for reviews, but if you have time to look at my code, i would appreciate.
I haven't put any "example" code (i have on local for the moment) but i guess this is better to push examples too
Tell me :)
C U

**Comment by pvcraven** [comment]
Thanks for the contribution and the revisions.

Took a first pass looking at it. The thing that bothers me with creating 'yet another' animated sprite is I've already burned through 4-5 versions of this concept. I hate to add more without it becoming a clear improvement. My last couple attempts I didn't add to the code base.

The only one we have that's actively used is for supporting animated tiles loaded from Tiled. They are simple and time based. Past that I'd recommend people try their own because I'm not confident of a good 'common' solution yet.

**Comment by Romuald78** [comment]
Raah ! So sad 😉
It was the first step to an evolution toward the use of TexturePacker, a tool i used with PhaserJS to easily integrate sprite animations
Anyway thx for the review 🙂

**Comment by pvcraven** [comment]
I'd keep at this, as a good solution would be great. Demoing it with some good proof-of-concept code, and maybe even a concept game would be a nice way to make sure that this is something that would survive the test of time.

**Comment by Romuald78** [comment]
Ok @pvcraven, do you mean it is worth for me to take time to create a real demo code , or that wont change much, and you dont plan to integrate it anyway
Excuse me for the question but my english level may be poor with some word/expression meanings

**Pull request closed without merging**